pkg/ebpf/c/bpf_bypass.h

```diff
@@ -12,40 +13,80 @@ BPF_ARRAY_MAP(program_bypassed, u32, 1)
     unsigned long bypass_program; \
     asm("%0 = " "bypass_program" " ll" : "=r"(bypass_program) :: "memory");
 
-#define BPF_BYPASSABLE_KPROBE(name, args...)					    \
-name(struct pt_regs *ctx);						    \
-static __always_inline typeof(name(0))					    \
-____##name(struct pt_regs *ctx, ##args);				    \
-typeof(name(0)) name(struct pt_regs *ctx)				    \
-{									    \
-    CHECK_BPF_PROGRAM_BYPASSED()                            \
-	_Pragma("GCC diagnostic push")					    \
-	_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")		    \
-	return ____##name(___bpf_kprobe_args(args));			    \
-	_Pragma("GCC diagnostic pop")					    \
-}									    \
-static __always_inline typeof(name(0))					    \
-____##name(struct pt_regs *ctx, ##args)
-
-#define BPF_BYPASSABLE_KRETPROBE(name, args...)					    \
-name(struct pt_regs *ctx);						    \
-static __always_inline typeof(name(0))					    \
-____##name(struct pt_regs *ctx, ##args);				    \
-typeof(name(0)) name(struct pt_regs *ctx)				    \
-{									    \
-    CHECK_BPF_PROGRAM_BYPASSED()                            \
-	_Pragma("GCC diagnostic push")					    \
-	_Pragma("GCC diagnostic ignored \"-Wint-conversion\"")		    \
-	return ____##name(___bpf_kretprobe_args(args));			    \
-	_Pragma("GCC diagnostic pop")					    \
-}									    \
-static __always_inline typeof(name(0)) ____##name(struct pt_regs *ctx, ##args)
+#define DO_BYPASS CHECK_BPF_PROGRAM_BYPASSED
+
+#define DO_WITH1(name, a) \
+static __always_inline void ____bpf_preamble__##name##__##a() { \
+    DO_##a() \
+}\
+
+#define DO_WITH2(name, a, b) \
+        DO_WITH1(name, a) \
+        DO_WITH1(name, b) \
+
+#define DO_WITH_CALL1(name, a) \
+        static __always_inline void ____bpf_preamble__##name() { \
+            ____bpf_preamble__##name##__##a(); \
+        }
+
+#define DO_WITH_CALL2(name, a, b) \
+        static __always_inline void ____bpf_preamble__##name() { \
+            ____bpf_preamble__##name##__##a(); \
+            ____bpf_preamble__##name##__##b(); \
+        }
+
+#define DO_WITHx(name, x,...) \
+        DO_WITH##x(name, __VA_ARGS__) \
+        DO_WITH_CALL##x(name, __VA_ARGS__)
+
+#define _WITH(name, x,...) \
+        DO_WITHx(name, x, __VA_ARGS__)
+
+#define WITH_PREAMBLE(name, ...) \
+        _WITH(name, nargs(__VA_ARGS__), __VA_ARGS__)
+
+#define WITH(...) \
+    __VA_ARGS__
+
+#define BPF_KPROBE_INSTR(preamble, name, args...) \
+    name(struct pt_regs *ctx);						    \
+    WITH_PREAMBLE(name, preamble) \
```

Comment:
Why define them beforehand as inline functions, and then call them in the body of the function? Couldn't we just add the code in the body like before?

Reply:
Reworked to avoid doing this. Calling preamble inline now